[MIG] stock_average_daily_sale: Backporting to 14.0 from 16.0

twalter-c2c opened the pull request.

rousseldenis committed:
[16.0][ADD] stock_average_daily_sale

rousseldenis committed:
[IMP] stock_average_daily_sale: Store the configurations on profile level + add demo

The configurations are now stored on abc profiles.

Some demo data are loaded in order to show this module features more easily.

rousseldenis committed:
[IMP] stock_average_daily_sale: Improve profile view

rousseldenis committed:
[IMP] Remove zone location dependency, rely on warehouse only

rousseldenis committed:
[IMP] stock_average_daily_sale: Improve tests

rousseldenis committed:
[IMP] stock_average_daily_sale: Improve module helps / change field name to recommended_qty

lmignon committed:
[FIX] stock_average_daily_sale: Don't exclude moves on priority

By default the priority in odoo is set to '0'. Before this change
the daily sale was never computed for moves done in a normal process.
There is no need to filter out moves based on the priority from
the computation.

rousseldenis committed:
[IMP] stock_average_daily_sale: Don't use Odoo demo data

sbejaoui committed:
[IMP] stock_average_daily_sale: add contributor

Co-authored-by: Jacques-Etienne Baudoux <je@bcim.be>

rousseldenis committed:
[IMP] stock_average_daily_sale: Avoid errors in all searches and return browse record

rousseldenis committed:
[FIX] stock_average_daily_sale: Use a new cursor to avoid closed one

rousseldenis committed:
[FIX] stock_average_daily_sale: Don't test the not refreshed materialized view

rousseldenis committed:
[FIX] stock_average_daily_sale: Don't check the view existence in tests

As tests create a Savepoint, there is a concurrent query when checking the
view existence.

rousseldenis committed:
[FIX] stock_average_daily_sale: Don't rely on tuples as id

As for x reason, some products appear more than one time in the report,
id generated by the concatenation of product_id and warehouse_id is
irrelevant (as duplicate values possible). Use row_number() instead

lmignon committed:
[IMP] stock_average_daily_sale: avoid concurrent update

Commit:
[UPD] Update stock_average_daily_sale.pot

OCA-git-bot committed:
[BOT] post-merge updates

mymage committed:
Added translation using Weblate (Italian)

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (58 of 58 strings)

Translation: stock-logistics-reporting-16.0/stock-logistics-reporting-16.0-stock_average_daily_sale
Translate-URL: https://translation.odoo-community.org/projects/stock-logistics-reporting-16-0/stock-logistics-reporting-16-0-stock_average_daily_sale/it/

twalter-c2c committed:
stock_average_daily_sale: Remove product_abc_classification dependency

Remove product_abc_classification dependency.
This module is not essential for the current module to function properly.

twalter-c2c committed:
stock_average_daily_sale: Change default access rights

Only stock_user can see config and report; stock_manager can create and delete config entries

twalter-c2c committed:
stock_average_daily_sale: Set default root location on WH

Set default average_daily_sale_root_location on WH create.
Default stock location of the new WH is used for this purpose

twalter-c2c committed:
stock_average_daily_sale: Allow to include/exclude weekends

Allow to decide if weekends should be included or excluded
in average daily sale calculation.

twalter-c2c committed:
stock_average_daily_sale: Take production location into account

Take also production location as dest location for daily usage calculation

twalter-c2c committed:
stock_average_daily_sale: Config adjustment - add from view and company_id

stock_average_daily_sale:
- add form view for config
- add company_id to config
- allow editing ABC classification level and make sure
  the level is unique per warehouse

twalter-c2c committed:
stock_average_daily_sale: Pre-migration script

twalter-c2c committed:
stock_average_daily_sale: Specify WH on demo stock moves

Commit:
[UPD] Update stock_average_daily_sale.pot

OCA-git-bot committed:
[BOT] post-merge updates

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: stock-logistics-reporting-16.0/stock-logistics-reporting-16.0-stock_average_daily_sale
Translate-URL: https://translation.odoo-community.org/projects/stock-logistics-reporting-16-0/stock-logistics-reporting-16-0-stock_average_daily_sale/

twalter-c2c committed:
[IMP] stock_average_daily_sale: black, isort, prettier

twalter-c2c committed:
[MIG] stock_average_daily_sale: Backporting to 14.0 from 16.0

simahawk commented:
@twalter-c2c we have a maturity level conflict
![image](https://github.com/user-attachments/assets/bc8b10d8-1847-4fd1-89ca-5037b67a9c6a)

See https://github.com/OCA/stock-logistics-warehouse/blob/14.0/stock_location_warehouse/__manifest__.py#L14
It can be raised to Beta if not Stable as is there since ages.

simahawk commented on stock_average_daily_sale/models/stock_warehouse.py:
```diff
@@ -29,11 +28,11 @@ def _compute_average_daily_sale_root_location_id(self):
                 continue
             warehouse.average_daily_sale_root_location_id = warehouse.lot_stock_id
 
-    @api.model_create_multi
```
why changing this?

twalter-c2c replied:
As discussed at the meeting: it reflects differences between 16.0 and 14.0 in the Odoo standard stock_warehouse create override.

simahawk commented on stock_average_daily_sale/tests/test_average_daily_sale.py:
```diff
@@ -188,10 +190,20 @@ def test_average_sale_profile_a(self):
             level="DEBUG",
-        ):
+        ) as cm:
+            logging.getLogger(
```
I'm not sure to understand this

twalter-c2c replied:
I added a comment explaining why I changed it and why in that way.

henrybackman commented on stock_average_daily_sale/migrations/16.0.1.1.0/pre-migrate.py:
Not sure if this is needed when backporting to 14 or should it be renamed to be `14.0.1.1.0` instead of the `16.0.1.1.0`

twalter-c2c replied:
You are right, the version is not correct, but anyway I think it's not needed when backporting. I removed the `migrations` from the module.

simahawk commented:
@rousseldenis would you mind to have a look? :pray:

sebalix commented on stock_average_daily_sale/tests/common.py:
```diff
@@ -0,0 +1,119 @@
+
+    @classmethod
+    def _create_inventory(cls):
+        inventory = cls.env["stock.inventory"].create(
```
```suggestion
        inventory = cls.inventory_obj.create(
```

sebalix commented on stock_average_daily_sale/tests/common.py:
```diff
@@ -50,20 +50,34 @@ def setUpClass(cls):
     @classmethod
     def _create_inventory(cls):
-        cls.inventory_obj.create(
+        inventory = cls.env["stock.inventory"].create(
```
Not sure why we switched from `stock.quant` to `stock.inventory` to put goods in stock, we can also use quant model to do that in v14. E.g:
```python
        cls.env["stock.quant"]._update_available_quantity(
            product,
            location,
            quantity,
            package_id=package,
            lot_id=lot,
            in_date=in_date,
        )
```
Or if we want to state the current qty in stock (not adding new ones), we can use this helper method:
```python
    @classmethod
    def _update_qty_in_location(
        cls, location, product, quantity, package=None, lot=None, in_date=None
    ):
        quants = cls.env["stock.quant"]._gather(
            product, location, lot_id=lot, package_id=package, strict=True
        )
        # this method adds the quantity to the current quantity, so remove it
        quantity -= sum(quants.mapped("quantity"))
        cls.env["stock.quant"]._update_available_quantity(
            product,
            location,
            quantity,
            package_id=package,
            lot_id=lot,
            in_date=in_date,
        )
```

But probably not blocking for a backport.

rousseldenis commented:
/ocabot migration stock_average_daily_sale

sebalix commented:
I found an issue, if there is no `warehouse_id` defined on stock moves (it's not required), the sql request doesn't provide any result. Fix ongoing.

EDIT: fixed

sebalix committed:
[FIX] stock_average_daily_sale: fallback on warehouse of src location if not defined on move

twalter-c2c committed:
[FIX] stock_average_daily_sale: include yesterday in calculation

Because of a datetime to date conversion, yesterday was excluded from calculation
while being set as a date_to value. With this commit we include all the stock
moves until up to 23:59:59 of the previous day.

twalter-c2c commented:
FTR: I noticed that previous day's moves were not being included in the calculations (although they were supposed to be). The last commit I posted a week ago fixes that issue.

jbaudoux commented:
/ocabot merge nobump

jbaudoux commented:
@twalter-c2c Can you port the 2 fixes to 16.0 ?

twalter-c2c commented:
> @twalter-c2c Can you port the 2 fixes to 16.0 ?

@jbaudoux Of course, I will do that next week. Thank you for merging the PR.
